### PR TITLE
Avoid client result invocation ID collisions

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes bullet 7 of https://github.com/dotnet/aspnetcore/issues/5280

Stephen had an interesting idea:
> Couldn't we pass in a per-connection, `IInvocationBinder` that adds the connection id prefix back to the invocation id so the `IHubProtocol` doesn't need to be aware of it?

Unfortunately, we don't know the connection ID when `IHubProtocol` queries `IInvocationBinder` so the invocation id has to be the entire source of truth for the lookup.

After doing some local perf testing, the best approach I found was to just generate a GUID and base64 encode it for every invocation. There is only one allocation, the final string, and GUID generation is still fairly cheap, the most expensive part is likely getting the random bytes.

Other alternatives considered:
* generate a shorter random prefix at startup and append an incrementing ID every invoke. Then every 1000 or so invocations we generate a new random prefix and reset the ID to 0 (so the full invocation ID doesn't grow too big). Sadly, I couldn't figure out how to do this lock free, or even lock free until it needed to reset. So this approach becomes very slow in a multi-threaded environment.
* generate a shorter random prefix at startup and append an incrementing ID every invoke. If a collision occurs, reply to the original server that it collided and original server will try to use a new ID, repeat. This approach would require more noise over the backplane and complexities with coding.
* Use some form of randomness and then just throw if a collision occurs. Collisions should be rare and application code likely already is handling exceptions because of timeouts, connection closing, connection throwing error for a result, etc.

### CODE REVIEW

**Comment by davidfowl** [comment]
Do you have any of the perf results? This is something we can change in the future if performance is a problem right?

File: src/SignalR/common/Shared/ClientResultsManager.cs [hunk 1]
```diff
@@ -42,6 +42,11 @@ public void AddInvocation(string invocationId, (Type Type, string ConnectionId,
     {
         var result = _pendingInvocations.TryAdd(invocationId, invocationInfo);
         Debug.Assert(result);
+        // Should have a 50% chance of happening once every 2.71 quintillion invocations (see UUID in Wikipedia)
```
Comment (davidfowl) [inline src/SignalR/common/Shared/ClientResultsManager.cs:45]: 🤣

File: src/SignalR/server/StackExchangeRedis/src/RedisHubLifetimeManager.cs [hunk 2]
```diff
@@ -784,6 +784,20 @@ private static string GenerateServerName()
...
+    {
+        Span<byte> buffer = stackalloc byte[16];
+        var success = Guid.NewGuid().TryWriteBytes(buffer);
+        Debug.Assert(success);
+        // 16 * 4/3 = 21.333 which means base64 encoding will use 22 characters of actual data and 2 characters of padding ('=')
+        Span<char> base64 = stackalloc char[24];
+        success = Convert.TryToBase64Chars(buffer, base64, out var written);
+        Debug.Assert(success);
+        Debug.Assert(written == 24);
+        // Trim the two '=='
```
Comment (davidfowl) [inline src/SignalR/server/StackExchangeRedis/src/RedisHubLifetimeManager.cs:797]:
```suggestion
        // Trim the two '=='
        Debug.Assert(base64.EndsWith("=="));
```

File: src/SignalR/server/StackExchangeRedis/src/RedisHubLifetimeManager.cs [hunk 3]
```diff
@@ -784,6 +784,20 @@ private static string GenerateServerName()
...
+        return new string(base64.Slice(0, base64.Length - 2));
```
Comment (davidfowl) [inline src/SignalR/server/StackExchangeRedis/src/RedisHubLifetimeManager.cs:798]: Fancy 😄 (probably less efficient though?)

```suggestion
        return new string(base64[..^2]);
```

Comment (BrennanConroy) [reply src/SignalR/server/StackExchangeRedis/src/RedisHubLifetimeManager.cs:798]: I dislike that syntax :/

Comment (davidfowl) [reply src/SignalR/server/StackExchangeRedis/src/RedisHubLifetimeManager.cs:798]: So much nicer

**Comment by BrennanConroy** [comment]
> This is something we can change in the future if performance is a problem right?

Yes, it's an internal implementation detail. Although if we wanted to do something co-operative it could be breaking if old server talks to new server.

100 threads, each thread doing 1,000,000 invocations:
| lock | new id every time | hardcoded "random" prefix and increment id every time |
|--|--|--|
| `21.3158182` | `2.9508009` | `3.6222541` |

**Review by davidfowl — APPROVED** [review]

**Comment by BrennanConroy** [comment]
@Pilchie for RC2. This fixes a bug when using multiple servers and the redis backplane in SignalR in combination with the new client result feature where it was possible for 2 IDs to collide which would cause the caller to hang in their app code until the cancellation token (they hopefully passed in) is canceled.

**Comment by Pilchie** [comment]
Approved for .NET 7 RC2.

**Pull request merged**